app/assets/javascripts/listings/ListingUnitService.js.coffee

```diff
@@ -29,49 +34,126 @@ ListingUnitService = ($http, ListingConstantsService, ListingIdentityService) ->
     else
       min
 
+  # FIXME: This might be broken by the new format
   Service.combineUnitSummaries = (listing) ->
-    # combined unitSummary is useful e.g. for overall occupancy levels across the whole listing
+    # Combined unitSummary is useful e.g. for overall occupancy levels across the whole listing.
     listing.unitSummaries ?= {}
     combined = _.concat(listing.unitSummaries.reserved, listing.unitSummaries.general)
     combined = _.omitBy(_.uniqBy(combined, 'unitType'), _.isNil)
-    # rename the unitType field to match how individual units are labeled
+    # Rename the unitType field to match how individual units are labeled.
     _.map(combined, (u) -> u.Unit_Type = u.unitType)
     Service._sortGroupedUnits(combined)
 
-  # Given ListingConstantsService.fieldsForUnitGrouping,
-  # combine similar units, and add up the number of available units
   Service._sumSimilarUnits = (units) ->
+    # Given ListingConstantsService.fieldsForUnitGrouping,
+    # combine similar units, and add up the number of available units.
     summaries = []
     # Create an identity function to group by unique price and income
     group = _.groupBy units, (unit) ->
       _.flatten(_.toPairs(_.pick(unit, ListingConstantsService.fieldsForUnitGrouping)))
     _.forEach group, (groupedUnits, id) ->
-      # Summarize each group by combining the unit details + # of units for that type/AMI combo
+      # Summarize each group by combining the unit details + # of units for that type/AMI combo.
       summary = _.pick(groupedUnits[0], ListingConstantsService.fieldsForUnitGrouping)
       summary.total = groupedUnits.length
       summaries.push(summary)
     summaries
 
-  # Group units by AMI % and Unit type
-  # Returns an object of the form:
-  #  {'50 (percent ami)':
-  #     {'1br': [{rent/price, # of units, etc}, ...], '2br': [...]},
-  #   '40 (percent ami)': {...}}
+  Service._getAMIAmount = (incomeList, occupancy) ->
+    if incomeList && incomeList.values?
+      _.find(incomeList['values'], {'numOfHousehold': occupancy})['amount']
+    else
+      console.error('Missing AMIs for occupancy', incomeList, occupancy)
+
+  Service._getIncomeRangesByOccupancy = (unitGroup) ->
+    # Given a row of units grouped by size, AMI, price, etc, determine the
+    # min and max incomes for that range for every available occupancy size.
+    # If a max occupancy isn't defined, as in the ownership case, default to showing up to 3
+    # occupancy options.
+    maxOccupancy = unitGroup.Max_Occupancy || unitGroup.Min_Occupancy + 2
+    occupancyRange = [unitGroup.Min_Occupancy..maxOccupancy]
+    # FIXME: This does not handle cases where listings have multiple AMI charts with the same percent.
```

Comment:
right now I think what we do in the max income table is just pick one of the charts if there is > 1 kind of chart with the same year + percent. It's pretty rare (I couldn't find any that are live now). 

Wondering if it makes sense to fix in this PR or do in a follow-up ticket.

Reply:
This PR is pretty big, I'd be happy with splitting it into a separate PR if that's not too much of a hassle